File: app/src/main/java/io/neoterm/setup/connection/NetworkConnection.java
package io.neoterm.setup.connection;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import io.neoterm.setup.SetupHelper;
import io.neoterm.setup.SourceConnection;

/**
 * @author kiva
 */

public class NetworkConnection implements SourceConnection {
    private final String sourceUrl;
    private HttpURLConnection connection = null;

    public NetworkConnection(String sourceUrl) {
        this.sourceUrl = sourceUrl;
    }

    @Override
    public InputStream getInputStream() throws IOException {
        if (connection == null) {
            connection = openHttpConnection();
            connection.setConnectTimeout(8000);
            connection.setReadTimeout(8000);
        }
        return connection.getInputStream();
    }

    @Override
    public int getSize() {
        if (connection != null) {
            return connection.getContentLength();
        }

        return 0;
    }

    @Override
    public void close() {
        if (connection != null) {
            connection.disconnect();
        }
    }

    private HttpURLConnection openHttpConnection() throws IOException {
        String arch = SetupHelper.INSTANCE.determineArchName();

        return (HttpURLConnection) new URL(sourceUrl + "/boot/" + arch + ".zip").openConnection();
    }
}

File: app/src/main/java/io/neoterm/setup/connections/NetworkConnection.kt
package io.neoterm.setup.connections

import io.neoterm.setup.SetupHelper
import io.neoterm.setup.SourceConnection
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL

/**
 * @author kiva
 */

class NetworkConnection(private val sourceUrl: String) : SourceConnection {
  private var connection: HttpURLConnection? = null

  @Throws(IOException::class)
  override fun getInputStream(): InputStream {
    if (connection == null) {
      connection = openHttpConnection()
      connection!!.connectTimeout = 8000
      connection!!.readTimeout = 8000
    }
    return connection!!.inputStream
  }

  override fun getSize(): Int {
    return if (connection != null) {
      connection!!.contentLength
    } else 0

  }

  override fun close() {
    if (connection != null) {
      connection!!.disconnect()
    }
  }

  @Throws(IOException::class)
  private fun openHttpConnection(): HttpURLConnection {
    val arch = SetupHelper.determineArchName()

    return URL("$sourceUrl/boot/$arch.zip").openConnection() as HttpURLConnection
  }
}
